intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/staticanalysis/IntelliJMarkupModelListener.java

```diff
@@ -0,0 +1,220 @@
+                .map(RangeHighlighterEx.class::cast);
+
+        final List<Warning<String>> warnings = rangeHighlighters
+                .filter(IntelliJMarkupModelListener::isWarningRangeHighlighter)
+                .map(rangeHighlighter -> createWarningFromRangeHighlighter(rangeHighlighter, null))
+                .map(IntelliJMarkupModelListener::classifyWarning)
+                .collect(Collectors.toList());
+
+        this.trackingEventManager.addEvent(new FileWarningSnapshotEvent(this.document.prepareDocument(), warnings));
+    }
+
+    private void flushForDocument() {
+        addCreatedWarnings(this.trackingEventManager, this.generatedWarnings.stream().map(IntelliJMarkupModelListener::classifyWarning), this.document);
+        this.generatedWarnings.clear();
+
+        addRemovedWarnings(this.trackingEventManager, this.warnings.stream().map(IntelliJMarkupModelListener::classifyWarning), this.document);
+        this.warnings.clear();
+    }
+
+    @Override
+    public void afterAdded(@NotNull RangeHighlighterEx rangeHighlighterEx) {
+        if (isWarningRangeHighlighter(rangeHighlighterEx)) {
+            final DateTime creationTime = DateTime.now();
+
+            this.timeMapping.put(rangeHighlighterEx, creationTime);
+            this.generatedWarnings.add(new Warning<>(rangeHighlighterEx, getLineNumberForHighlighter(rangeHighlighterEx), creationTime.toDate()));
+        }
+    }
+
+    @Override
+    public void beforeRemoved(@NotNull RangeHighlighterEx rangeHighlighterEx) {
+        DateTime creationTime = this.timeMapping.remove(rangeHighlighterEx);
+
+        if (isWarningRangeHighlighter(rangeHighlighterEx)) {
+            // Only process a deletion if we hadn't encountered this marker in this session before.
+            // If we did encounter it, remove returns `true` and the warning is not saved as removed.
+            if (!this.generatedWarnings.removeIf(warning -> warning.type.equals(rangeHighlighterEx))) {
+                this.warnings.add(createWarningFromRangeHighlighter(rangeHighlighterEx, creationTime));
+            }
+        }
+    }
+
+    @Override
+    public void attributesChanged(@NotNull RangeHighlighterEx rangeHighlighterEx, boolean b, boolean b1) {
+    }
+
+    @Override
+    public void dispose() {
+    }
+
+    private static boolean isWarningRangeHighlighter(@NotNull RangeHighlighterEx rangeHighlighterEx) {
+        return rangeHighlighterEx.getLayer() == HighlighterLayer.WARNING;
+    }
+
+    @NotNull
+    private Warning<RangeHighlighterEx> createWarningFromRangeHighlighter(@NotNull RangeHighlighterEx rangeHighlighterEx, DateTime creationTime) {
+        final DateTime now = DateTime.now();
+
+        return new Warning<>(
+                rangeHighlighterEx,
+                getLineNumberForHighlighter(rangeHighlighterEx),
+                now.toDate(),
+                creationTime == null ? -1 : Seconds.secondsBetween(creationTime, now).getSeconds()
```

Comment:
We avoid the ? operator in WatchDog

Reply:
This does not seem to be documented anywhere in the wiki. Would be great if such rules could be added there, as I was unaware of this rule.